deps/rabbit/src/rabbit_exchange.erl

```diff
@@ -286,49 +247,33 @@ lookup_scratch(Name, App) ->
 -spec update_scratch(name(), atom(), fun((any()) -> any())) -> 'ok'.
 
 update_scratch(Name, App, Fun) ->
-    rabbit_misc:execute_mnesia_transaction(
-      fun() ->
-              _ = update(Name,
-                     fun(X = #exchange{scratches = Scratches0}) ->
-                             Scratches1 = case Scratches0 of
-                                              undefined -> orddict:new();
-                                              _         -> Scratches0
-                                          end,
-                             Scratch = case orddict:find(App, Scratches1) of
-                                           {ok, S} -> S;
-                                           error   -> undefined
-                                       end,
-                             Scratches2 = orddict:store(
-                                            App, Fun(Scratch), Scratches1),
-                             X#exchange{scratches = Scratches2}
-                     end),
-              ok
-      end).
-
--spec update_decorators(name()) -> 'ok'.
-
-update_decorators(Name) ->
-    rabbit_misc:execute_mnesia_transaction(
-      fun() ->
-              case mnesia:wread({rabbit_exchange, Name}) of
-                  [X] -> _ = store_ram(X),
-                         ok;
-                  []  -> ok
-              end
-      end).
-
--spec update
-        (name(),
-         fun((rabbit_types:exchange()) -> rabbit_types:exchange()))
-         -> not_found | rabbit_types:exchange().
-
-update(Name, Fun) ->
-    case mnesia:wread({rabbit_exchange, Name}) of
-        [X] -> X1 = Fun(X),
-               store(X1);
-        []  -> not_found
+    Decorators = case rabbit_db_exchange:get(Name) of
+                     {ok, X} -> rabbit_exchange_decorator:active(X);
+                     {error, not_found} -> []
+                 end,
```

Comment:
I can't find where this comes from either in the original code. Do you remember where?

Reply:
It doesn't come exactly as that. The previous `rabbit_exchange:store` would set the decorators for the ram record here: https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbit/src/rabbit_exchange.erl#L157 In prep work for Khepri, we collect the decorators outside of the transaction and pass them to the update/store functions. Decorators can't be collected inside a Khepri function because they use apply to call modules not known in compile time. If you consider this is too ahead of time, I can revert it to the old behaviour of calculating decorators inside of the transaction.

Reply:
I see. No need to revert, it's probably ok this way and as you said, we can't keep the current behavior in a Khepri world.